File: python3/695-max-area-of-island.py
from collections import deque
from typing import List


def maxAreaOfIsland(grid: List[List[int]]) -> int:
    m = len(grid)
    n = len(grid[0])
    max_area = 0

    def bfs_visit(x, y) -> int:
        bfs_queue = deque()
        seen = set()

        # init
        bfs_queue.append((x,y))
        seen.add((x,y))
        grid[x][y] = 0
        count = 1

        while bfs_queue:
            node = bfs_queue.popleft()

            directions = [(-1,0), (1,0), (0,-1), (0,1)]
            for d_i, d_j in directions:
                new_i = node[0] + d_i
                new_j = node[1] + d_j

                if (0 <= new_i < m and 0 <= new_j < n):
                    # adjacent nodes
                    if ((new_i, new_j) not in seen) and (grid[new_i][new_j] == 1):
                        seen.add((new_i, new_j))
                        grid[new_i][new_j] = 0
                        bfs_queue.append((new_i, new_j))
                        count += 1

        return count

    for i in range(m):
        for j in range(n):
            if grid[i][j] == 1:
                max_area = max(max_area, bfs_visit(i,j))

    return max_area

grid = [[0,0,1,0,0,0,0,1,0,0,0,0,0],[0,0,0,0,0,0,0,1,1,1,0,0,0],[0,1,1,0,1,0,0,0,0,0,0,0,0],[0,1,0,0,1,1,0,0,1,0,1,0,0],[0,1,0,0,1,1,0,0,1,1,1,0,0],[0,0,0,0,0,0,0,0,0,0,1,0,0],[0,0,0,0,0,0,0,1,1,1,0,0,0],[0,0,0,0,0,0,0,1,1,0,0,0,0]]
print(maxAreaOfIsland(grid))

File: cpp/695-max-area-of-island.cpp
#include <iostream>
#include <print>
#include <vector>
#include <queue>
using namespace std;

class Solution
{
public:
    int getArea(int i, int j, int M, int N, vector<vector<int>> &grid)
    {
        // floodfill algorithm
        vector<pair<int, int>> directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        queue<pair<int, int>> q;

        // init
        q.push({i, j});
        grid[i][j] = 0;
        int size = 0;

        while (!q.empty())
        {
            auto curr = q.front();
            q.pop();
            size++;

            for (int i = 0; i < 4; i++)
            {
                int adj_i = curr.first + directions[i].first;
                int adj_j = curr.second + directions[i].second;

                if (0 > adj_i || adj_i >= M || 0 > adj_j || adj_j >= N)
                {
                    continue;
                }
                if (grid[adj_i][adj_j] == 0)
                {
                    continue;
                }
                q.push({adj_i, adj_j});
                grid[adj_i][adj_j] = 0;
            }
        }

        return size;
    }
    int maxAreaOfIsland(vector<vector<int>> &grid)
    {
        int M = grid.size();
        int N = grid[0].size();
        int max_area = 0;
        for (int i = 0; i < M; i++)
        {
            for (int j = 0; j < N; j++)
            {
                if (grid[i][j] == 1)
                {
                    max_area = max(max_area, getArea(i, j, M, N, grid));
                }
            }
        }
        return max_area;
    }
};

int main()
{
    Solution solution;
    vector<vector<int>> grid = {
        {0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0},
        {0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0},
        {0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0},
        {0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0},
        {0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0},
        {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0},
        {0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0},
        {0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0}};
    cout << solution.maxAreaOfIsland(grid) << endl;
    return 0;
}